[AST] Don't merge memory locations in AliasSetTracker

brunodf-gf opened the pull request:
This changes the AliasSetTracker to track memory locations instead of pointers in its alias sets. The motivation for this is outlined in [an RFC posted on LLVM discourse](https://discourse.llvm.org/t/rfc-dont-merge-memory-locations-in-aliassettracker/73336).

This commit does not yet change method names and code comments to refer to "memory location(s)" instead of "pointer(s)", to keep the changeset focused, and to first gain feedback on the idea. An additional commit can take care of this.

In the data structures of the AST implementation, I made the choice to replace the linked list of `PointerRec` entries (that had to go anyway) with a simple flat vector of `MemoryLocation` objects, but for the `AliasSet` objects referenced from a lookup table, I retained the mechanism of a linked list, reference counting, forwarding, etc. The data structures could be revised in a follow-up change.

brunodf-snps commented on polly/lib/Analysis/ScopBuilder.cpp:
Possibly it makes more sense to retrieve a `MemoryAccess` through a `MemoryLocation` rather than a pointer.

brunodf-snps commented:
Tagging some possible reviewers:
r? @nikic 
r? @jdoerfert
r? @efriedma-quic

jdoerfert commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -40,36 +40,37 @@ static cl::opt<unsigned>
+bool AliasSet::isMustAliasMergeWith(AliasSet &AS, BatchAAResults &BatchAA) const {
+  for (const MemoryLocation &MemLoc : MemoryLocs)
+    for (const MemoryLocation &ASMemLoc : AS.MemoryLocs)
+      if (!BatchAA.isMustAlias(MemLoc, ASMemLoc))
```
Must-alias is not a transitive relationship? Before we picked a representative and only checked those, why do we need to check all entries now?

brunodf-snps replied:
This is discussed in the RFC, using a representative also implies merging of the LocationSize, AAInfo,... of the other accesses into it: https://discourse.llvm.org/t/rfc-dont-merge-memory-locations-in-aliassettracker/73336#mustalias-sets-saturation-4

brunodf-snps replied:
Even though a representative is not possible, @nikic convinced me that any must-alias pair is sufficient because must-alias is transitive. See commit 5f82129f98f2.

jdoerfert commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -551,7 +472,7 @@ AliasSet &AliasSetTracker::addPointer(MemoryLocation Loc,
   AS.Access |= E;
 
-  if (!AliasAnyAS && (TotalMayAliasSetSize > SaturationThreshold)) {
+  if (!AliasAnyAS && (TotalAliasSetSize > SaturationThreshold)) {
```
Unclear if we should increase the threshold now. I guess we can wait and see.

jdoerfert reviewed (commented):
I left comments. Generally, I think this is the right way to go. We had the custom rolled stuff that was expanded from MemLoc, and used to create MemLoc all the time anyway. 

Do we have compile time data? Maybe like this and w/ a SetVector instead of a plain vector?

brunodf-snps commented:
Thanks @jdoerfert for your comments. There is compile-time data here: https://llvm-compile-time-tracker.com/compare.php?from=fab25949688d7b038ee047d4e8b44a23803db3bd&to=aa7ccadde93f78148026103a03c282e23acbf0b3&stat=instructions:u

I will address your other comments.

alinas commented:
The proposal SGTM!
I was somewhat surprised this doesn't affect compile time at all, with it doing AA checks on all MemoryLocation pairs, but this seems to indeed not have any impact (tested on internal benchmarks too). My assumption is that most alias checks are taking advantage of the cache in BatchAA.

I'll follow up with inline comments.
+1 to update method names to match the change to tracking memory locations, and update comments.

nikic commented on llvm/include/llvm/Analysis/AliasSetTracker.h:
```diff
@@ -307,7 +173,7 @@ class AliasSetTracker {
   ilist<AliasSet> AliasSets;
 
-  using PointerMapType = DenseMap<AssertingVH<Value>, AliasSet::PointerRec *>;
+  using PointerMapType = DenseMap<MemoryLocation, AliasSet *>;
```
I don't really get why this needs to be indexed by MemoryLocation. I'd still expect all MemoryLocations for the same Value to be in the same AliasSet, so it seems like the existing indexing should work fine...

brunodf-snps replied:
> I'd still expect all MemoryLocations for the same Value to be in the same AliasSet

I was unsure if this always holds. In case of full restrict support, I think the same pointer value is used for the two stores in the following snippet:

```
int * restrict r = p;
*p = 123;
*r = 456;
```

But they have different provenance and metadata, and ScopedNoAliasAA will report that they do not alias. But this snippet has undefined behavior (because of course, the accesses do alias).

I guess that BasicAA will quickly return MustAlias for a query with the same pointer Value regardless of LocationSize, metadata, etc.? Then indeed another AA can only say NoAlias in an illegal situation.

brunodf-snps replied:
> I don't really get why this needs to be indexed by MemoryLocation [...] it seems like the existing indexing should work fine...

I still tried this, see [this branch](https://github.com/brunodf-snps/llvm-project/tree/ast-pointer-map). The pointer map can be smaller, but this is traded for extra work in `AliasSetTracker::getAliasSetFor`, see commit 23443f3. It turns out that two memory locations for the same `UndefValue` **do** end up in different alias sets (see [this BasicAA logic](https://github.com/llvm/llvm-project/blob/92fc4b482f58cb243d982d572adb7a80ceab6945/llvm/lib/Analysis/BasicAliasAnalysis.cpp#L1472-L1475)) so I have to exclude them from the pointer map, which complicate the logic. On the plus side, a lots of tests can be reverted, see commit 92f01a2, since we can restore the "pointer values" output in the alias set dump, but I don't know if that is much of an argument.

Let me know if you consider the version on the branch an improvement, then I will include it in this pull request.

brunodf-snps replied:
I still realized there was a problem with commit 23443f3. When a memory location with an undef pointer value is added to an alias set, that alias set is not referenced from the pointer map entry of that pointer value (because the pointer value may appear in _multiple_ alias sets). But without the reference to the alias set from some access structure, the reference count of the set may reach zero, and the alias set may be dropped. (For similar reason, the reference count of an alias set is increased by 1 when at least one unknown instruction is registered in it.)

I solved this problem (on [the aforementioned branch](https://github.com/brunodf-snps/llvm-project/commits/ast-pointer-map/)) by keeping a collection with alias sets that contain undef pointer values. This keeps those alias sets referenced, so they are not dropped. It also allows to scan only those alias sets, when we are trying to find if a memory location with an undef pointer value is already registered or not.

nikic replied:
I'm confused. Yes, AA treats undef as noalias, but as far as I know AST specifically wants the undefs to be part of the same set, or at least that is how I interpret the comment at https://github.com/llvm/llvm-project/blob/5f254eb05566f5b400a212ae77117a9efd9019a1/llvm/lib/Analysis/AliasSetTracker.cpp#L346-L350. It seems like you are going out of your way to have undefs in different sets -- why?

More generally I would expect that all MemLocs that have the same pointer will end up in the same alias set, even if AA reports them as NoAlias (as implied by metadata or undef). After all they are *also* MustAlias, and treating them as such makes things a lot simpler.

brunodf-snps replied:
I agree that when using a map from pointer values, it is more logical to simply place MemLocs in an existing set of the pointer value, even if AA reports NoAlias for this set. I've updated the ast-pointer-map branch to this effect, in particular, see:
https://github.com/brunodf-snps/llvm-project/blob/9e9ebcd53aff011cff2b4d7ebdf439855806894b/llvm/lib/Analysis/AliasSetTracker.cpp#L312-L319

But be aware that in a program where AA indicates that two instruction using an undef pointer are not aliasing, _putting them in the same alias set is information loss_ (although it may not be relevant):

```
%val = load i32, ptr undef
...
store ptr null, ptr undef
``` 

So I would disagree with the cited code comment, or at least find it misleading:

>> Since alias(undef, undef) is NoAlias, mergeAliasSetsForPointer(undef, ...) will not find the the right set for undef, even if it exists. 

`mergeAliasSetsForPointer` is correctly indicating (with a `nullptr` result) that there is no set that the second memory location aliases with, according to AA. The existing set for the first memory location with an undef pointer is only a "right" set insofar that the implementation with a pointer map cannot accurately represent the situation. (But it may be a fine tradeoff to be inaccurate here.)

> It seems like you are going out of your way to have undefs in different sets -- why?

Well, I have roughly 2 objectives for AST with this patch: (1) avoid information loss and (2) minimal knowledge about AA or MemoryLocation internals.  So I am reluctant to add information loss for an implementation organization (the pointer map). But you seem to favor the pointer map for conceptual reasons too, so then I would propose what is on my ast-pointer-map branch.

brunodf-snps replied:
It is true that both NoAlias and MustAlias are correct results here and you can choose which one to pick, but my reasoning was that AA already picks NoAlias and so I attempted to represent the same thing in AST (when I say I strive for no information loss, I essentially mean that AST should construct alias sets consistent with AA results -- the motivating example is a case where AA is indicating NoAlias, and current AST is ending up with MayAlias due to implementation organization).

But I am fine with an implementation where all MemoryLocations for the same pointer Value are by construction in the same AliasSet, i.e. where the PointerMap stays. The discussion has shown that the impact is only for undef pointer values, and I expect the difference is of very little practical relevance (as you basically write, creating a new set for every undef pointer does not seem to have added value). So I have merged my ast-pointer-map branch into this PR now, so the combined set of changes that I propose is visible here, and we can move forward with that.

brunodf-snps commented on llvm/lib/Transforms/Scalar/LoopVersioningLICM.cpp:
```diff
@@ -258,15 +258,15 @@ bool LoopVersioningLICM::legalLoopMemoryAccesses() {
+      const Value *Ptr = A.Ptr;
       // Alias tracker should have pointers of same data type.
-      TypeCheck = (TypeCheck && (SomePtr->getType() == Ptr->getType()));
+      TypeCheck = (TypeCheck && (SomePtr->getType() == Ptr->getType())); // FIXME: no longer effective check since opaque pointers?
```
Done. In my rebase, I split this off as commit 951157ac9861. If OK, could you precommit this on main on my behalf?

brunodf-snps commented on llvm/test/Transforms/LICM/variant-aainfo.ll:
```diff
@@ -0,0 +1,65 @@
+; yet the LICM failed to prote load/store ptr %s out of the loop
+
+; RUN: opt < %s -passes=aa-eval -evaluate-aa-metadata -aa-pipeline=basic-aa,tbaa -print-all-alias-modref-info -disable-output 2>&1 | FileCheck --check-prefix=CHECKAA %s
+; RUN: opt < %s -S -passes=licm | FileCheck --check-prefix=CHECKLICM %s
```
Done. In my rebase, I split of a first version of the test with "fixme" regarding the missed optimization as commit 0bab78b35175. If OK, could you precommit this on main on my behalf?

brunodf-snps replied:
New version of the test intended for precommit (after addressing the comments from @nikic) in commit 7fd5d3cd328d.

brunodf-snps commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -671,7 +671,7 @@ class AccessAnalysis {
   void addStore(MemoryLocation &Loc, Type *AccessTy) {
     Value *Ptr = const_cast<Value*>(Loc.Ptr);
-    AST.add(Ptr, LocationSize::beforeOrAfterPointer(), Loc.AATags);
+    AST.add(Loc.getWithNewSize(LocationSize::beforeOrAfterPointer()));
```
In my rebase, I split off the change to switch to `Loc.getWithNewSize` as commit 91668131a016. If OK, could you precommit this on main on my behalf?

brunodf-snps replied:
Actually, I think we can precommit this a little bit wider, and already update all callers of this particular add method. I've reviewed my branch history and the new commit that I'd propose to precommit is 048677aead99 (this replaces the commit mentioned above).

brunodf-snps replied:
Commit rebased as 071f0b9f40a7 (but exactly the same diff).

nikic commented on llvm/test/Transforms/LICM/variant-aainfo.ll:
```diff
@@ -0,0 +1,66 @@
+; RUN: opt < %s -passes=aa-eval -evaluate-aa-metadata -aa-pipeline=basic-aa,tbaa -print-all-alias-modref-info -disable-output 2>&1 | FileCheck --check-prefix=CHECKAA %s
```
Remove this RUN line.

brunodf-snps replied:
Fixed in commit 7fd5d3cd328d.

nikic commented on llvm/test/Transforms/LICM/variant-aainfo.ll:
```diff
@@ -0,0 +1,66 @@
+; RUN: opt < %s -passes=aa-eval -evaluate-aa-metadata -aa-pipeline=basic-aa,tbaa -print-all-alias-modref-info -disable-output 2>&1 | FileCheck --check-prefix=CHECKAA %s
+; RUN: opt < %s -S -passes=licm | FileCheck --check-prefix=CHECKLICM %s
```
Use update_test_checks.py.

brunodf-snps replied:
I will look into to it, but out of curiosity: why do you mandate generated assertions? From https://llvm.org/docs/TestingGuide.html#generating-assertions-in-regression-tests I understood generated assertions are optional, if too tedious to write manually. Is this to update the test in the future?

nikic replied:
See https://llvm.org/docs/TestingGuide.html#best-practices-for-regression-tests for basic testing guidelines.

brunodf-snps replied:
Fixed in commit 7fd5d3cd328d.

nikic commented on llvm/test/Transforms/LICM/variant-aainfo.ll:
```diff
@@ -0,0 +1,66 @@
+!14 = !{!9, !9, i64 0}
+!15 = distinct !{!15, !16, !17}
+!16 = !{!"llvm.loop.mustprogress"}
+!17 = !{!"llvm.loop.unroll.disable"}
```
Remove all unnecessary metadata. Only keep the tbaa metadata.

brunodf-snps replied:
Fixed in commit 7fd5d3cd328d.

nikic commented on llvm/test/Transforms/LICM/variant-aainfo.ll:
```diff
@@ -0,0 +1,66 @@
+; CHECKLICM: store i32 {{%[^ ]+}}, ptr [[P]]
+; CHECKLICM: [[P_NEXT]] = getelementptr inbounds i32, ptr [[P]], i32 1
+
+define void @_Z4testP1S(ptr noundef %s) #0 {
```
Drop attributes.

brunodf-snps replied:
Fixed in commit 7fd5d3cd328d.

nikic reviewed (commented):
(Just some comments on the test, before it can be pre-committed.)

brunodf-gf committed:
AliasSetTracker: don't merge memory locations

nikic commented:
I have landed https://github.com/llvm/llvm-project/commit/358e7656807f7a459baa0ad6002dd246fbad88e3, https://github.com/llvm/llvm-project/commit/c0307789796150699b550a314be236ea0b1be9ed and https://github.com/llvm/llvm-project/commit/3c9236c0bf57a3a7689fc64c57c1325ea466b1c0. Note that I slightly modified the second commit to drop the extra wrapper method.

brunodf-snps commented:
> I have landed [358e765](https://github.com/llvm/llvm-project/commit/358e7656807f7a459baa0ad6002dd246fbad88e3), [c030778](https://github.com/llvm/llvm-project/commit/c0307789796150699b550a314be236ea0b1be9ed) and [3c9236c](https://github.com/llvm/llvm-project/commit/3c9236c0bf57a3a7689fc64c57c1325ea466b1c0). Note that I slightly modified the second commit to drop the extra wrapper method.

Thanks! I have rebased again to make the precommitted changes disappear from this PR.

brunodf-gf committed:
Document isMustAliasMerge method

brunodf-gf committed:
Introduce getPointers method

For clients that want to iterate the deduplicated pointers. Also
consistently distinguish memory location vs. pointer iteration.

brunodf-gf committed:
Switch back to map of pointer values

brunodf-gf committed:
Update tests

brunodf-snps commented on llvm/include/llvm/Analysis/AliasSetTracker.h:
```diff
@@ -49,99 +49,12 @@ class Value;
   AliasSet *Forward = nullptr;
 
+  /// Memory locations in this alias set.
+  std::vector<MemoryLocation> MemoryLocs;
```
I added a convenience method `getPointers` (cf. your other remark regarding a convenience API) for clients that relied on this. This calculates the set of deduplicated pointers at the moment it is called. I think maintaining the set in addition to the memory locations is probably not worth it, considering the use cases that I see in the tree.

brunodf-snps commented on polly/lib/Analysis/ScopDetectionDiagnostic.cpp:
```diff
@@ -639,7 +639,8 @@ bool ReportNonSimpleMemoryAccess::classof(const RejectReason *RR) {
   for (const auto &I : AS)
-    Pointers.push_back(I.getValue());
+    if (!llvm::is_contained(Pointers, I.Ptr))
+      Pointers.push_back(I.Ptr);
```
Done.

brunodf-snps commented on polly/lib/Analysis/ScopDetection.cpp:
```diff
@@ -1158,8 +1158,9 @@ bool ScopDetection::isValidAccess(Instruction *Inst, const SCEV *AF,
 
-        for (const auto &Ptr : AS) {
-          Instruction *Inst = dyn_cast<Instruction>(Ptr.getValue());
+        for (const auto &MemLoc : AS) {
```
This was added as `getPointers`.

brunodf-snps commented on polly/lib/Analysis/ScopBuilder.cpp:
```diff
@@ -3238,8 +3238,10 @@ ScopBuilder::buildAliasGroupsForAccesses() {
+    llvm::SmallPtrSet<const Value *, 8> Seen;
+    for (auto &MemLoc : AS)
+      if (Seen.insert(MemLoc.Ptr).second)
+        AG.push_back(PtrToAcc[const_cast<Value *>(MemLoc.Ptr)]);
```
Done.

brunodf-snps commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -1100,7 +1100,7 @@ bool AccessAnalysis::canCheckPtrAtRT(RuntimePointerChecking &RtCheck,
     SmallVector<MemAccessInfo, 4> AccessInfos;
     for (const auto &A : AS) {
-      Value *Ptr = A.getValue();
+      Value *Ptr = const_cast<Value *>(A.Ptr);
```
I'm not happy about it either, but I don't immediately see how this can be improved. Conceptually, analyses can use `const Value *`. `MemoryLocation` and AA is doing this, but other analyses are not: that leads to some const-casts (also before this change `LoopAccessAnalysis` had some of these casts). Another source of const-casts is that a transformation retrieves `const Value *` from an analysis, and casts away the const qualification to work on them (that is fine, if the values fed into the analysis were non-const to begin with).

brunodf-snps commented on llvm/include/llvm/Analysis/AliasSetTracker.h:
```diff
@@ -284,16 +150,16 @@ class AliasSet : public ilist_node<AliasSet> {
-  void addPointer(AliasSetTracker &AST, PointerRec &Entry, LocationSize Size,
-                  const AAMDNodes &AAInfo, bool KnownMustAlias = false,
-                  bool SkipSizeUpdate = false);
+  bool isMustAliasMergeWith(AliasSet &AS, BatchAAResults &BatchAA) const;
```
Documentation was added. Actually, the auxiliary function was only introduced to exit elegantly from a nested loop. I still removed the function from the class, and made it internal to the AliasSetTracker.cpp file.

brunodf-snps replied:
> Actually, the auxiliary function was only introduced to exit elegantly from a nested loop.

This was perhaps unnecessarily heavy. I still removed the function, and replaced it by an immediately-invoked lambda, so the logic is directly visible in `AliasSet::mergeSetIn`.

brunodf-snps commented on llvm/include/llvm/Analysis/AliasSetTracker.h:
```diff
@@ -207,66 +118,21 @@ class AliasSet : public ilist_node<AliasSet> {
-  bool empty() const { return PtrList == nullptr; }
+  using iterator = std::vector<MemoryLocation>::const_iterator;
+  iterator begin() const { return MemoryLocs.begin(); }
+  iterator end()   const { return MemoryLocs.end(); }
```
The source of the const-casts is not the const-iterator, it is [inside `MemoryLocation`](https://github.com/llvm/llvm-project/blob/617ee255cab59598600f453f9e80091aa203324b/llvm/include/llvm/Analysis/MemoryLocation.h#L230).

I think it would be possible to provide a convenience version of `getPointers` that integrates the const-cast so the client would not be burdened with this, but the caller would still be responsible to ensure the cast is valid. Is there a precedent of this in LLVM? I'm only aware of methods that intregrate const-casts when we know it is valid ([example](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Analysis/ValueTracking.h#L681-L685)).

brunodf-gf committed:
getPointers: logical to reserve for Seen set too

brunodf-gf committed:
Replace isMustAliasMerge by immediately invoked lambda

brunodf-gf committed:
Keep alias sets with undef pointer values in UndefPointerSets

This allows to scan only these sets when we want to check if a memory
locations was already registered, but importantly, it also keeps these
alias sets referenced.

brunodf-snps commented:
Apart from updating the method names and comments (which is the reason why I still keep the PR marked as draft), I think I have now addressed all of the comments up to this point.

nikic commented on llvm/include/llvm/Analysis/AliasSetTracker.h:
```diff
@@ -49,99 +50,12 @@ class Value;
   AliasSet *Forward = nullptr;
 
+  /// Memory locations in this alias set.
+  std::vector<MemoryLocation> MemoryLocs;
```
```suggestion
  SmallVector<MemoryLocation, 0> MemoryLocs;
```

brunodf-snps replied:
Fixed in commit efc787bc5d61.

nikic commented on llvm/lib/Analysis/LoopAccessAnalysis.cpp:
```diff
@@ -1140,10 +1141,11 @@ bool AccessAnalysis::canCheckPtrAtRT(RuntimePointerChecking &RtCheck,
-                       MemAccessInfo AccessWrite(AC.getValue(), true);
+      assert((ASPointers.size() <= 1 ||
+              all_of(ASPointers,
+                     [this](auto Ptr) {
```
```suggestion
                     [this](const Value *Ptr) {
```

brunodf-snps replied:
Fixed in commit efc787bc5d61.

nikic commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -46,30 +46,29 @@ void AliasSet::mergeSetIn(AliasSet &AS, AliasSetTracker &AST,
+    [&] {
+      for (const MemoryLocation &MemLoc : *this)
+        for (const MemoryLocation &ASMemLoc : AS)
+          if (!BatchAA.isMustAlias(MemLoc, ASMemLoc)) {
```
Wouldn't it be sufficient to check that *any* pair is MustAlias?

brunodf-snps replied:
OK, I think I was too strict. The [definition of MustAlias](https://github.com/llvm/llvm-project/blob/061e4f24b24a3b59d73a94dc6f2f0d21a2b7beac/llvm/docs/AliasAnalysis.rst?plain=1#L138-L140) seems indeed transitive, so any must-alias pair is sufficient (even if the alias queries do not necessarily return must-alias for all pairs).

Updated in commit 5f82129f98f2.

nikic commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -124,42 +110,20 @@ void AliasSet::removeFromTracker(AliasSetTracker &AST) {
-  Entry.updateSizeAndAAInfo(Size, AAInfo);
+  if (isMustAlias() && !KnownMustAlias)
+    for (const MemoryLocation &ASMemLoc : MemoryLocs)
+      if (!AST.getAliasAnalysis().isMustAlias(ASMemLoc, MemLoc)) {
```
Here again, isn't it sufficient if we MustAlias with any one location?

brunodf-snps replied:
Updated in commit 5f82129f98f2.

nikic commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -246,42 +191,54 @@ ModRefInfo AliasSet::aliasesUnknownInst(const Instruction *Inst,
+  PointerVector Result;
+  if (MemoryLocs.size() <= Result.capacity()) {
+    for (const MemoryLocation &MemLoc : MemoryLocs)
+      if (llvm::find(Result, MemLoc.Ptr) == Result.end())
```
is_contained

brunodf-snps replied:
Code removed in commit efc787bc5d61.

nikic commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -246,42 +191,54 @@ ModRefInfo AliasSet::aliasesUnknownInst(const Instruction *Inst,
+AliasSet::PointerVector AliasSet::getPointers() const {
+  // To deduplicate pointer values, use a linear scan if the number of elements
+  // is small, or a set if large. This is the same idea as SmallSetVector. In
+  // addition, we can allocate space for the result vector upfront.
```
Why not use SmallSetVector directly? You can use `takeVector()` at the end if you don't want to keep SmallVector as the return type.

brunodf-snps replied:
The reason not use `SmallSetVector` was to do the allocations upfront, using `MemoryLocs.size()` as a good estimate and upper bound of the size. But that was probably overthinking it. I now switched to `SmallSetVector` with `takeVector()` as suggested (commit efc787bc5d61). If `SmallSetVector` ever gains a `reserve` that could still be used here.

brunodf-gf committed:
Fix small remarks

brunodf-gf committed:
Sufficient to check any pair for MustAlias; assumed to be transitive

brunodf-gf committed:
getAliasSetFor: replace logic for undef values

Simply place them in the earlier alias set that was known for them,
although this entails information loss.

brunodf-gf committed:
Merge branch 'ast-pointer-map' into ast-dont-merge-memory-locations

brunodf-snps commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -317,62 +262,67 @@ AliasSet *AliasSetTracker::findAliasSetForUnknownInst(Instruction *Inst) {
-    return *AliasAnyAS;
+    AS = AliasAnyAS;
+  } else if (AliasSet *AliasAS =
+                 mergeAliasSetsForPointer(MemLoc, MustAliasAll)) {
```
No, because adding a new memory location may merge existing alias sets.

Suppose we first add memory location (%p, 2) (notation meaning a memory location with pointer value %p and location size 2). Then we add memory location (%p+2, 2). These memory locations are not aliasing and will be in different alias sets. Then we add memory location (%p, 4). The map entry for %p is pointing to the alias set of (%p, 2) but the memory location is also aliasing with the alias set of (%p+2, 2). mergeAliasSetsForPointer is needed to merge the alias sets, and we add the new memory location to the merged set that it returns.

In the process of merging, one of the original alias sets will become empty and forward to the other. A bit lower in the code, we do verify that the map entry, after following possible forwarding, must end up in the same merged alias set.

nikic replied:
Ah yes, you're right.

brunodf-snps commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -317,62 +262,67 @@ AliasSet *AliasSetTracker::findAliasSetForUnknownInst(Instruction *Inst) {
+    // (This is only known to occur for undef pointer values, which AA treats as
+    // noalias.)
+    AS = MapEntry->getForwardedTarget(*this);
+    AS->Alias = AliasSet::SetMayAlias;
```
At this point, AA has indicated that the new memory location is no-alias with any set, including the set from the map entry. If that set was a must-alias set, it would seem that we have to downgrade to may-alias when adding the new memory location there to be consistent with the no-alias result?

I suppose your reasoning is that any two memory locations with the same pointer value are always must-alias (even if AA indicates they are also no-alias), so we know that the new memory location is also must-alias with the earlier memory location in the set for which map entry was added, and by transitivity with all memory locations in the set?

nikic replied:
Yes, exactly. I doubt this really matters in practice, but I think keeping the MustAlias here should be safe (and very marginally simpler...)

brunodf-snps replied:
OK. But then I think it is better to fully build in that the we assume a must-alias result for the existing alias set with a memory location using the same pointer value, and use this in `mergeAliasSetsForPointer`. I've pushed an updated version, see new commits b7fbb33320a9 and ad4fa84fd279. (This is maybe what you were after with your earlier remark about not calling `mergeAliasSetsForPointer` when there is a MapEntry: while I think we must call the method and do merging, we can save some alias queries there.)

nikic replied:
Thanks, your updated version looks very clean!

nikic reviewed (commented):
This looks good toe me.

brunodf-gf committed:
Directly assume alias query with existing map entry is a MustAlias

brunodf-snps commented on llvm/lib/Analysis/AliasSetTracker.cpp:
```diff
@@ -317,62 +262,67 @@ AliasSet *AliasSetTracker::findAliasSetForUnknownInst(Instruction *Inst) {
+  AliasSet *&MapEntry = PointerMap[MemLoc.Ptr];
+  if (MapEntry) {
+    AliasSet *AS = MapEntry->getForwardedTarget(*this);
+    if (llvm::is_contained(AS->MemoryLocs, MemLoc)) {
```
I don't mind, but then shouldn't I also remove the prefix from `llvm::any_of`, `llvm::append_range`, etc. in my changes to this file? What's the common practice here?

nikic replied:
Yes, you should not use `llvm::` in llvm/.

brunodf-snps replied:
OK, done in commit 68f3016.

brunodf-gf committed:
Remove llvm:: prefix in changes

brunodf-gf committed:
Simplify definition and usage of getForwardedTarget

Integrate this private method with updating the map entry in unified
procedure.

brunodf-snps commented:
> This looks good to me.

Thanks! I will still do a pass over the method names and comments to use "memory location" terminology.

@jdoerfert I was hoping to have this landed before the llvm-18 branch is cut (Jan 23, plus account for some days margin). Do you still have remarks?

brunodf-gf committed:
Remove headers that have become unused

brunodf-gf committed:
Update terminology in comments and method names

Change "pointer(s)" to "memory location(s)" or "element(s)" as
appropriate.

brunodf-gf committed:
Change terminology in alias set printing

nikic reviewed (approved):
LGTM

brunodf-snps commented:
Thanks for committing. main has a new test since my branch point llvm/test/Analysis/BasicAA/separate_storage-alias-sets.ll
that is failing, let me fix that.

brunodf-snps commented:
I pushed a fix for the failing test on the original branch of this PR:
https://github.com/brunodf-snps/llvm-project/commit/052119475489be8092e6d332d09b6d694d7dd217

brunodf-snps commented:
See fix in commit https://github.com/brunodf-snps/llvm-project/commit/f90637de685fadf17da7426b4765936ee7fd3991 on branch https://github.com/brunodf-snps/llvm-project/commits/ast-fix-new-test/

nikic commented:
@brunodf-snps Thanks, committed in https://github.com/llvm/llvm-project/commit/509f634f76caa40d522875e17188ca77df9159b5.
